File: go/kv/wrap_kv.go
package kv

import (
	"github.com/mabels/runified/types"
)

type wrappedKvStore[T any] struct {
	blobStore types.KeyValueStoreBlob
	serde     types.SerDe
	factory   func() *T
}

// Del implements types.KeyValueStore.
func (w *wrappedKvStore[T]) Del(key string) (*T, error) {
	my := w.factory()
	bytes, err := w.blobStore.Del(key)
	if err != nil {
		return my, err
	}
	w.serde.Unmarshal(bytes, &my)
	return my, err
}

// Get implements types.KeyValueStore.
func (w *wrappedKvStore[T]) Get(key string) (*T, error) {
	bytes, err := w.blobStore.Get(key)
	if err != nil {
		return nil, err
	}
	my := w.factory()
	err = w.serde.Unmarshal(bytes, my)
	return my, err
}

// Set implements types.KeyValueStore.
func (w *wrappedKvStore[T]) Set(key string, value *T) error {
	bytes, err := w.serde.Marshal(value)
	if err != nil {
		return err
	}
	return w.blobStore.Set(key, bytes)
}

func WrapKvStore[T any](blobStore types.KeyValueStoreBlob, serde types.SerDe, factory func() *T) types.KeyValueStore[T] {
	return &wrappedKvStore[T]{
		blobStore: blobStore,
		serde:     serde,
		factory:   factory,
	}
}

File: ts/src/kv/wrap_kv.ts
import { KeyValueStore, KeyValueStoreBlob } from "../types/key_value";
import { SerDe } from "../types/serde";

interface wrappedKvStoreParams<TT> {
  readonly blobStore: KeyValueStoreBlob;
  readonly serde: SerDe<TT>;
}

class wrappedKvStore<T> implements KeyValueStore<T> {
  readonly _blobStore: KeyValueStoreBlob;
  readonly _serde: SerDe<T>;
  // readonly _factory: (t: never) => Result<T, Error>;

  constructor(args: wrappedKvStoreParams<T>) {
    this._blobStore = args.blobStore;
    this._serde = args.serde;
    // this._factory = args.factory;
  }

  // Del implements types.KeyValueStore.
  async Del(key: string): Promise<T> {
    try {
      const bytes = await this._blobStore.Del(key);
      const res = this._serde.Unmarshal(bytes);
      if (res.is_ok()) {
        return Promise.resolve(res.unwrap());
      }
      return Promise.reject(res.unwrap_err());
    } catch (err) {
      return Promise.reject(err);
    }
  }

  // Get implements types.KeyValueStore.
  async Get(key: string): Promise<T> {
    try {
      const bytes = await this._blobStore.Get(key);
      const res = this._serde.Unmarshal(bytes);
      if (res.is_ok()) {
        return Promise.resolve(res.unwrap());
      }
      return Promise.reject(res.unwrap_err());
    } catch (err) {
      return Promise.reject(err);
    }
  }

  // Set implements types.KeyValueStore.
  async Set(key: string, value: T): Promise<T> {
    const bytes = this._serde.Marshal(value);
    if (bytes.is_err()) {
      return Promise.reject(bytes.unwrap_err());
    }
    await this._blobStore.Set(key, bytes.unwrap());
    return Promise.resolve(value);
  }
}

export function WrapKvStore<T>(
  blobStore: KeyValueStoreBlob,
  serde: SerDe<T>,
  // factory: (t: any) => Result<T, Error>
): KeyValueStore<T> {
  return new wrappedKvStore<T>({
    blobStore: blobStore,
    serde: serde,
  });
}
